src/verifier/step3/Step3Logic.sol

```diff
@@ -0,0 +1,359 @@
+        elementsToHash[counter] = U2.comm_W.x;
+        counter++;
+        elementsToHash[counter] = U2.comm_W.y;
+        counter++;
+        if (Pallas.isInfinity(U2.comm_W)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        // Absorb X
+        for (uint256 i = 0; i < U2.X.length; i++) {
+            elementsToHash[counter] = U2.X[i];
+            counter++;
+        }
+
+        // Absorb comm_T
+        Pallas.PallasAffinePoint memory comm_T = Pallas.decompress(nifs.compressed_comm_T);
+        elementsToHash[counter] = comm_T.x;
+        counter++;
+        elementsToHash[counter] = comm_T.y;
+        counter++;
+        if (Pallas.isInfinity(comm_T)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        require(counter == NUM_FE_FOR_RO); 
+
+        // uint32 absorbLen = uint32(counter);
+        // uint32 squeezeLen = 1;
+        // uint32 domainSeparator = 0;
+
+        SpongeOpLib.SpongeOp memory absorb = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Absorb, uint32(counter));
+        SpongeOpLib.SpongeOp memory squeeze = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Squeeze, 1);
+        SpongeOpLib.SpongeOp[] memory pattern = new SpongeOpLib.SpongeOp[](2);
+        pattern[0] = absorb;
+        pattern[1] = squeeze;
+        IOPatternLib.IOPattern memory p = IOPatternLib.IOPattern(pattern);
+        
+        NovaSpongePallasLib.SpongeU24Pallas memory sponge = NovaSpongePallasLib.start(p, 0);
+
+        sponge = NovaSpongePallasLib.absorb(sponge, elementsToHash);
+
+        (, uint256[] memory output) = NovaSpongePallasLib.squeeze(sponge, 1);
+        sponge = NovaSpongePallasLib.finishNoFinalIOCounterCheck(sponge);
+
+        // uint256 r = output[0] & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff;
+
+        RelaxedR1CSInstance memory result = foldInstance(U1, U2, comm_T, output[0] & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff);
+
+        return result;
+    }
+
+    function foldInstance(
+        RelaxedR1CSInstance calldata U1,
+        R1CSInstance calldata u2,
+        Pallas.PallasAffinePoint memory comm_T,
+        uint256 r
+    ) public view returns (RelaxedR1CSInstance memory) {
+        // uint256[] memory x1 = U1.X;
+        // Pallas.PallasAffinePoint memory comm_W_1 = U1.comm_W;
+        // Pallas.PallasAffinePoint memory comm_E_1 = U1.comm_E;
+        // uint256 u1 = U1.u;
+
+        uint256[] memory x2 = u2.X;
+        Pallas.PallasAffinePoint memory comm_W_2 = u2.comm_W;
+
+        require(U1.X.length == x2.length, "Witness vectors do not match length");
+
+        uint256[] memory X = new uint256[](U1.X.length);
+
+        for (uint256 i = 0; i < x2.length; i++) {
+            X[i] = addmod(U1.X[i], mulmod(r, x2[i], Pallas.R_MOD), Pallas.R_MOD);
```

Comment:
Why do we use `R_MOD`, not `P_MOD`? I remember that multiplications and additions seem to require P_MOD, isn't it?

Reply:
I think the simple answer is that it's what makes the functions pass with the test vector... I presume this is because the witness vector X is encoded in terms of the scalar field rather than the base field...

Reply:
This is the equivalent line in Nova:
https://github.com/microsoft/Nova/blob/1e6bf942e215979b0e02d5b2b83a6c4450d740e1/src/r1cs.rs#L519
The inputs are indeed Scalars for the concerned group, which means that the correct modulus is Pallas.R_MOD (== Vesta.P_MOD).